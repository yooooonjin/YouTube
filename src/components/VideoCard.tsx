import { Link } from 'react-router-dom';
import { Video } from '../model/video';
type Props = {
  video: Video;
  direction?: 'col' | 'row';
};
export default function VideoCard({ video, direction = 'col' }: Props) {
  return (
    <Link
      to={`/videos/watch/${video.id}`}
      className={`${direction === 'row' && 'flex gap-4'}`}
    >
      <div className='min-w-40'>
        <img
          src={video.thumbnail}
          alt='video_thumbnale'
          className='w-full aspect-video object-cover'
        />
      </div>
      <div className='text-xs'>
        <p className='line-clamp-2 mb-2'>{video.title}</p>
        <p className='text-[0.6rem] font-light'>{video.channelTitle}</p>
        <p className='text-[0.6rem] font-light'>
          {video.publishedAt.getDate()} days ago
        </p>
      </div>
    </Link>
  );
}
