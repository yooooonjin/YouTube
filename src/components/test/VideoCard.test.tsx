import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route, useLocation } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withRouter } from '../../test/utils';
import { fakeVideo as video } from '../../test/videos';
import VideoCard from '../VideoCard';

describe('VideoCart', () => {
  const { channelTitle, thumbnail, title, id } = video;

  it('renders col type correctly', () => {
    const component = renderer.create(
      withRouter(<Route path='/' element={<VideoCard video={video} />} />)
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders row type correctly', () => {
    const component = renderer.create(
      withRouter(
        <Route path='/' element={<VideoCard video={video} direction='row' />} />
      )
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders video item', () => {
    //Given
    render(
      withRouter(<Route path='/' element={<VideoCard video={video} />} />)
    );

    //When
    const image = screen.getByRole('img') as HTMLImageElement;
    expect(image.src).toBe(thumbnail);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
  });

  it('navigates to detailed video page when clicked', async () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }
    render(
      withRouter(
        <>
          <Route path='/' element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );

    const card = screen.getByRole('link');
    await userEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
