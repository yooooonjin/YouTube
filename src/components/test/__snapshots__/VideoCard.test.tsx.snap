// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCart renders col type correctly 1`] = `
<a
  className="false"
  href="/videos/watch/video id"
  onClick={[Function]}
>
  <div
    className="min-w-40"
  >
    <img
      alt="video title"
      className="w-full aspect-video object-cover"
      src="http://image/"
    />
  </div>
  <div
    className="text-xs"
  >
    <p
      className="line-clamp-2 mb-2"
    >
      video title
    </p>
    <p
      className="text-[0.6rem] font-light"
    >
      video channel
    </p>
    <p
      className="text-[0.6rem] font-light"
    >
      4
       days ago
    </p>
  </div>
</a>
`;

exports[`VideoCart renders row type correctly 1`] = `
<a
  className="flex gap-4"
  href="/videos/watch/video id"
  onClick={[Function]}
>
  <div
    className="min-w-40"
  >
    <img
      alt="video title"
      className="w-full aspect-video object-cover"
      src="http://image/"
    />
  </div>
  <div
    className="text-xs"
  >
    <p
      className="line-clamp-2 mb-2"
    >
      video title
    </p>
    <p
      className="text-[0.6rem] font-light"
    >
      video channel
    </p>
    <p
      className="text-[0.6rem] font-light"
    >
      4
       days ago
    </p>
  </div>
</a>
`;
